src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py

```diff
@@ -0,0 +1,441 @@
+
+    x, y = point
+
+    x_bytes = int(x).to_bytes(64, "big")
+    y_bytes = int(y).to_bytes(64, "big")
+
+    return x_bytes + y_bytes
+
+
+def decode_G1_scalar_pair(data: Bytes) -> Tuple[Point2D, int]:
+    """
+    Decode 160 bytes to a G1 point and a scalar.
+
+    Parameters
+    ----------
+    data :
+        The bytes data to decode.
+
+    Returns
+    -------
+    point : Tuple[Point2D, int]
+        The G1 point and the scalar.
+
+    Raises
+    ------
+    InvalidParameter
+        If the sub-group check failed.
+    """
+    assert len(data) == 160
+    p = bytes_to_G1(buffer_read(data, U256(0), U256(128)))
+    if multiply(p, curve_order) is not None:
+        raise InvalidParameter("Sub-group check failed.")
+
+    m = int.from_bytes(buffer_read(data, U256(128), U256(32)), "big")
+
+    return p, m
+
+
+def bytes_to_FQ(
+    data: Bytes, optimized: bool = False
+) -> Union[FQ, OPTIMIZED_FQ]:
+    """
+    Decode 64 bytes to a FQ element.
+
+    Parameters
+    ----------
+    data :
+        The bytes data to decode.
+    optimized :
+        Whether to use the optimized FQ implementation.
+
+    Returns
+    -------
+    fq : Union[FQ, OPTIMIZED_FQ]
+        The FQ element.
+
+    Raises
+    ------
+    InvalidParameter
+        If the field element is invalid.
+    """
+    assert len(data) == 64
+    c = int.from_bytes(data[:64], "big")
+
+    if c >= P:
+        raise InvalidParameter("Invalid field element")
+
+    if optimized:
+        return OPTIMIZED_FQ(c)
+    else:
+        return FQ(c)
+
+
+def bytes_to_FQ2(
+    data: Bytes, optimized: bool = False
+) -> Union[FQ2, OPTIMIZED_FQ2]:
+    """
+    Decode 128 bytes to a FQ2 element.
+
+    Parameters
+    ----------
+    data :
+        The bytes data to decode.
+    optimized :
+        Whether to use the optimized FQ2 implementation.
+
+    Returns
+    -------
+    fq2 : Union[FQ2, OPTIMIZED_FQ2]
+        The FQ2 element.
+
+    Raises
+    ------
+    InvalidParameter
+        If the field element is invalid.
+    """
+    assert len(data) == 128
+    c_0 = int.from_bytes(data[:64], "big")
+    c_1 = int.from_bytes(data[64:], "big")
+
+    if c_0 >= P:
+        raise InvalidParameter("Invalid field element")
+    if c_1 >= P:
+        raise InvalidParameter("Invalid field element")
+
+    if optimized:
+        return OPTIMIZED_FQ2((c_0, c_1))
+    else:
+        return FQ2((c_0, c_1))
+
+
+def bytes_to_G2(data: Bytes) -> Point2D:
+    """
+    Decode 256 bytes to a G2 point. Does not perform sub-group check.
+
+    Parameters
+    ----------
+    data :
+        The bytes data to decode.
+
+    Returns
+    -------
+    point : Point2D
+        The G2 point.
+
+    Raises
+    ------
+    InvalidParameter
+        Either a field element is invalid or the point is not on the curve.
+    """
+    assert len(data) == 256
+
+    x = bytes_to_FQ2(data[:128])
+    y = bytes_to_FQ2(data[128:])
+
+    assert isinstance(x, FQ2) and isinstance(y, FQ2)
```

Comment:
When wouldn't this be the case? Is `OPTIMIZED_FQ2` an `FQ2`?

Reply:
`OPTIMIZED_FQ2` has been defined in `py_ecc` optimize mapping the field_element to curve. As such, the assertion was put in place to prevent `mypy` from complaining about the Union.